BLE_GattServer_ExperimentalServices/source/main.cpp

```diff
@@ -0,0 +1,179 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2006-2015 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "ble/BLE.h"
+#include "ble/gap/Gap.h"
+#include "mbed-os-ble-utils/pretty_printer.h"
+#include "mbed-os-experimental-ble-services/services/inc/LinkLossService.h"
+
+using namespace std::literals::chrono_literals;
+
+const static char DEVICE_NAME[] = "LinkLoss";
+
+const static char  MILD_ALERT[] = "Mild Alert!";
+const static char  HIGH_ALERT[] = "High Alert!";
+
+static events::EventQueue event_queue(/* event count */ 10 * EVENTS_EVENT_SIZE);
+static ChainableGapEventHandler chainable_gap_event_handler;
+
+class LinkLossDemo : ble::Gap::EventHandler, LinkLossService::EventHandler {
+public:
+    LinkLossDemo(BLE &ble, events::EventQueue &event_queue, ChainableGapEventHandler &chainable_gap_event_handler) :
+            _ble(ble),
+            _event_queue(event_queue),
+            _chainable_gap_event_handler(chainable_gap_event_handler),
+            _link_loss_service(_ble, _event_queue, _chainable_gap_event_handler),
+            _adv_data_builder(_adv_buffer)
+    {
+    }
+
+    void start()
+    {
+        _ble.init(this, &LinkLossDemo::on_init_complete);
+
+        _event_queue.dispatch_forever();
+    }
+
+private:
+    void on_init_complete(BLE::InitializationCompleteCallbackContext *params)
+    {
+        if (params->error != BLE_ERROR_NONE) {
+            printf("Ble initialization failed.");
+            return;
+        }
+
+        print_mac_address();
+
+        _chainable_gap_event_handler.addEventHandler(this);
```

Comment:
you should add some comments explaining what a chainable event handler is and how it works

Reply:
Yes please explain that two parts of the application requires to handle events from Gap. To do this we use ChainableGapEventHandler which will dispatch events from gap to all its registered handlers.